packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java

```diff
@@ -128,28 +128,26 @@ public void modifyIndexHtmlResponse(IndexHtmlResponse response) {
     }
 
     protected Map<String, AvailableViewInfo> collectClientViews(
-            Predicate<? super String> isUserInRole,
-            boolean isUserAuthenticated) {
+            VaadinRequest request) {
+
         if (!deploymentConfiguration.isProductionMode()) {
             clientRouteRegistry.loadLatestDevModeFileRoutesJsonIfNeeded(
                     deploymentConfiguration);
         }
 
-        return clientRouteRegistry.getAllRoutes().values().stream()
-                .filter(config -> config.getRouteParameters() == null
-                        || config.getRouteParameters().isEmpty()
-                        || config.getRouteParameters().values().stream()
-                                .noneMatch(
-                                        param -> param == RouteParamType.REQUIRED))
-                .filter(config -> routeUtil.isRouteAllowed(isUserInRole,
-                        isUserAuthenticated, config))
-                .map(config -> new AvailableViewInfo(config.getTitle(),
-                        config.getRolesAllowed(), config.isLoginRequired(),
-                        config.getRoute(), config.isLazy(),
-                        config.isAutoRegistered(), config.menu(),
-                        Collections.emptyList(), config.getRouteParameters()))
-                .collect(Collectors.toMap(AvailableViewInfo::route,
-                        Function.identity()));
+        return MenuRegistry
+                .collectClientMenuItems(true, deploymentConfiguration, request)
+                .entrySet().stream()
+                .filter(view -> !hasRequiredParameter(
+                        view.getValue().routeParameters()))
+                .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
```

Comment:
I think it worth checking that the `MenuRegistry#validateViewAccessible` (used in vaadin/flow#19389) is actually checking A&A in the same way that `filter(config -> routeUtil.isRouteAllowed(isUserInRole, isUserAuthenticated, config))` was doing recursively for the views and the possible parent layout(s) configurations.

Reply:
The MenuRegistry removes all child routes for a route that is not valid, which is the same as done here, but from parent to child instead of child to parent.

The output should be exactly the same with no changes in the result.